### PR TITLE
typescript declarations emit

### PR DESCRIPTION
**Pull request by Bnaya**
Emit typescript declarations - utilize typescript 3.7 ability to emit types for js code.
Branched out from #3805 

Minor code changes to make the types compilation pass

### CODE REVIEW

File: packages/react-admin/src/index.ts [hunk 1]
```diff
@@ -5,5 +5,11 @@ import AdminRouter from './AdminRouter';
 import defaultI18nProvider from './defaultI18nProvider';
 
 export * from 'ra-core';
+/* 
+error - what do to? which one is the correct symbol?
+Module 'ra-core' has already exported a member named 'Notification'. Consider explicitly re-exporting to resolve the ambiguity.ts(2308)
+Module 'ra-core' has already exported a member named 'Pagination'. Consider explicitly re-exporting to resolve the ambiguity.ts(2308)
+*/
+// @ts-ignore
```
Comment (Bnaya) [inline packages/react-admin/src/index.ts]: @fzaninotto what's the correct symbols here should be?

**Comment by user753** [comment]
There is many required props like
`classes` required
```
declare function Datagrid({ classes: classesOverride, ...props }: {
    [x: string]: any;
    classes: any;
}): JSX.Element;
```
`record` required
```
declare function ReferenceField({ children, record, source, ...props }: {
    [x: string]: any;
    children: any;
    record: any;
    source: any;
}): JSX.Element;
```
ReferenceInput 
all required
```
interface Props {
    allowEmpty: boolean;
    basePath: string;
    children: ReactElement;
    classes: any;
    className: string;
    label: string;
    reference: string;
    resource: string;
    [key: string]: any;
}
```

**Comment by Bnaya** [comment]
@user753 this is by no-mean full typescript coverage.
The types from the js files are based on the guesses of TSC.
But it is a great start for symbol discovery and a way to make incremental improvements to the types,
via tsx refactor or jsdocs-types in exiting js files

**Comment by user753** [comment]
@Bnaya I understand that. I think It coulb be fixed if all guessed props were optional. 
Maybe there is a simple way to do that? 
Also if these types was on https://github.com/DefinitelyTyped/DefinitelyTyped we could easily fix them.

File: packages/ra-ui-materialui/src/layout/Loading.js [hunk 2]
```diff
@@ -33,6 +33,9 @@ const useStyles = makeStyles(
     { name: 'RaLoading' }
 );
 
+/**
+ * @type {React.FunctionComponent<{className: string; classes: object; loadingPrimary?: string; loadingSecondary?: string;}>}
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/layout/Loading.js]: I prefer that you turn that component to TypeScript. I don't want to maintain a lib with 3 different syntax (JS, TS, and JS-with-types)

File: packages/react-admin/src/index.ts (hunk 1 above)
Comment (fzaninotto) [reply packages/react-admin/src/index.ts]: I believe ra-ui-material-ui should import those types from ra-core

**Comment by fzaninotto** [comment]
Also, before emitting the type definitions from ra-ui-material-ui, we must migrate the demo to TypeScript, so that we can make sure the types don't break existing apps. Would you like to tackle that task first?

**Comment by Bnaya** [comment]
> migrate the demo ...Would you like to tackle that task first

I've enabled checkJs for the demo and ill go over the errors i'm getting, That's should be equivalent to typescript refactor +-
I've also deleted the `examples/demo/src/react-admin.d.ts` file, and now my VSCODE picking up the types built from my workspace.

In general, i have the feeling the component composition api RA works with not playing well with strict typescript. for example:
Screen Shot 2020-02-03 at 18 32 11
That's a valid RA code, but there is no way to tell typescript that "if your parent component has source prop you don't need one of your own"
So i'm not sure to where to take it from here

> that's because we want to be sure that the types we publish don't break existing TS apps

I think it will be very difficult to guarantee that no typescript app will not break in any way after publishing types with `react-admin`
each project has it's own hacks around types now
For me it's `@ts-ignore` all over the imports, for others its empty react admin `d.ts`
Even if our new types would be perfect, it might collide with any `d.ts` that in the consuming projects

**Comment by fzaninotto** [comment]
On the example you underline, if the source can be injected, in that case it should not be required in the types.

**Comment by Bnaya** [comment]
I'm willing to put effort to make that happen, but there are quite few things that are not clear to me
I would like to talk directly if possible to find the best course of action

**Comment by Vadorequest** [comment]
@Bnaya Check this out https://github.com/marmelab/react-admin/issues/1617#issuecomment-595874113 :)

**Comment by fzaninotto** [comment]
@Bnaya What isn't clear and what do you want to talk about?

**Comment by matepaiva** [comment]
I am really looking forward to this feature! It's really helpful!

**Comment by falkenhawk** [comment]
@fzaninotto since you closed this, I guess this is now superceded by https://github.com/marmelab/react-admin/issues/4505 ?

**Pull request closed without merging**